datafusion/optimizer/src/reduce_outer_join.rs

```diff
@@ -0,0 +1,355 @@
+                _optimizer,
+                &join.right,
+                nonnullable_cols,
+                _optimizer_config,
+            )?;
+
+            Ok(LogicalPlan::Join(Join {
+                left: Arc::new(left_plan),
+                right: Arc::new(right_plan),
+                join_type: new_join_type,
+                join_constraint: join.join_constraint,
+                on: join.on.clone(),
+                filter: join.filter.clone(),
+                schema: join.schema.clone(),
+                null_equals_null: join.null_equals_null,
+            }))
+        }
+        LogicalPlan::Projection(Projection {
+            input,
+            expr,
+            schema,
+            alias: _,
+        }) => {
+            let projection = schema
+                .fields()
+                .iter()
+                .enumerate()
+                .map(|(i, field)| {
+                    // strip alias, as they should not be part of filters
+                    let expr = match &expr[i] {
+                        Expr::Alias(expr, _) => expr.as_ref().clone(),
+                        expr => expr.clone(),
+                    };
+
+                    (field.qualified_name(), expr)
+                })
+                .collect::<HashMap<_, _>>();
+
+            // re-write all Columns based on this projection
+            for col in nonnullable_cols.iter_mut() {
+                if let Some(Expr::Column(column)) = projection.get(&col.flat_name()) {
+                    *col = column.clone();
+                }
+            }
+
+            // optimize inner
+            let new_input = reduce_outer_join(
+                _optimizer,
+                input,
+                nonnullable_cols,
+                _optimizer_config,
+            )?;
+
+            from_plan(plan, expr, &[new_input])
+        }
+        _ => {
+            let expr = plan.expressions();
+
+            // apply the optimization to all inputs of the plan
+            let inputs = plan.inputs();
+            let new_inputs = inputs
+                .iter()
+                .map(|plan| {
+                    reduce_outer_join(
+                        _optimizer,
+                        plan,
+                        nonnullable_cols,
+                        _optimizer_config,
+                    )
+                })
+                .collect::<Result<Vec<_>>>()?;
+
+            from_plan(plan, &expr, &new_inputs)
+        }
+    }
+}
+
+/// Recursively traversese expr, if expr returns false when
+/// any inputs are null, treats columns of both sides as nonnullable columns.
+///
+/// For and/or expr, extracts from all sub exprs and merges the columns.
+/// For or expr, if one of sub exprs returns true, discards all columns from or expr.
+/// For IS NOT NULL/NOT expr, always returns false for NULL input.
+///     extracts columns from these exprs.
+/// For all other exprs, fall through
+fn extract_nonnullable_columns(
```

Comment:
What do you think about using `Expr::nullable` instead?  That seems very similar / the same as what you are doing with this function (trying to determine if an expression can be nullable / non nullable)? 

https://github.com/apache/arrow-datafusion/blob/d985c0a3b6d96b02028f3abb6edb361ea72cac14/datafusion/expr/src/expr_schema.rs#L148-L211

If `Expr::nullable` doesn't work correctly I think we should update that

Reply:
The extract_nonnullable_columns gets all columns of these exprs which returns false if inputs are null， and we use these columns to see if they appear in null rows of output for outer join.  If so, this means these null rows does not meet the conditions, we can filter null rows, reduce outer to inner. So I think the purpose of extract_nonnullable_columns is not the same as Expr::nullable.